include/verilated_random.cpp

```diff
@@ -307,27 +290,31 @@ bool VlRandomVar::set(std::string&& val) const {
     return true;
 }
 
-std::shared_ptr<const VlRandomExpr> VlRandomizer::randomConstraint(VlRNG& rngr, int bits) {
-    unsigned long long hash = VL_RANDOM_RNG_I(rngr) & ((1 << bits) - 1);
-    std::shared_ptr<const VlRandomExpr> concat = nullptr;
-    std::vector<std::shared_ptr<const VlRandomExpr>> varbits;
-    for (const auto& var : m_vars) {
-        for (int i = 0; i < var.second->width(); i++)
-            varbits.emplace_back(std::make_shared<const VlRandomExtract>(var.second, i));
-    }
+void VlRandomizer::randomConstraint(std::ostream& os, VlRNG& rngr, int bits) {
+    const IData hash = VL_RANDOM_RNG_I(rngr) & ((1 << bits) - 1);
+    int varBits = 0;
+    for (const auto& var : m_vars) varBits += var.second->width();
+    os << "(= #b";
+    for (int i = bits - 1; i >= 0; i--) os << (VL_BITISSET_I(hash, i) ? '1' : '0');
+    if (bits > 1) os << " (concat";
     for (int i = 0; i < bits; i++) {
-        std::shared_ptr<const VlRandomExpr> bit = nullptr;
-        for (unsigned j = 0; j * 2 < varbits.size(); j++) {
-            unsigned idx = j + VL_RANDOM_RNG_I(rngr) % (varbits.size() - j);
-            auto sel = varbits[idx];
-            std::swap(varbits[idx], varbits[j]);
-            bit = bit == nullptr ? sel : std::make_shared<const VlRandomBinOp>("bvxor", bit, sel);
+        IData varBitsLeft = varBits;
+        IData varBitsWant = (varBits + 1) / 2;
```

Comment:
```suggestion
        const IData varBitsLeft = varBits;
        const IData varBitsWant = (varBits + 1) / 2;
```

Reply:
These are not const on purpose, as they are both decremented in the loop.